kbncontent.go

```diff
@@ -0,0 +1,248 @@
+		return "", nil
+	}
+
+	if title, err := jsonpath.Get("$.attributes.title", doc); err == nil {
+		return title.(string), nil
+	}
+
+	if title, err := jsonpath.Get("$.title", doc); err == nil {
+		return title.(string), nil
+	}
+
+	// by-value dashboard case
+	if title, err := jsonpath.Get("$.embeddableConfig.savedVis.title", doc); err == nil {
+		return title.(string), nil
+	}
+
+	return "", nil
+}
+
+// Attaches domain knowledge as well as information from within the document
+func attachMetaInfo(desc *VisDesc) {
+	if result, err := getVisType(desc.Doc, desc.SoType); err == nil {
+		desc.Type = result
+	}
+
+	if result, err := getVisTitle(desc.Doc, desc.SoType); err == nil {
+		desc.Title = result
+	}
+
+	if result, err := getTSVBType(desc.Doc, desc.Type); err == nil {
+		desc.TSVBType = result
+	}
+
+	if isTSVB(desc.Type) {
+		desc.SemanticType = desc.TSVBType
+	} else {
+		desc.SemanticType = desc.Type
+	}
+
+	desc.IsLegacy = isLegacy(desc.SoType, desc.Type)
+}
+
+func deserializeSubPaths(doc objx.Map) (map[string]interface{}, error) {
+	uiState := doc.Get("attributes.uiStateJSON")
+	if uiState.IsStr() {
+		if parsed, err := objx.FromJSON(uiState.Str()); err == nil {
+			doc.Set("attributes.uiStateJSON", parsed)
+		}
+	}
+
+	visState := doc.Get("attributes.visState")
+	if visState.IsStr() {
+		if parsed, err := objx.FromJSON(visState.Str()); err == nil {
+			doc.Set("attributes.visState", parsed)
+		}
+	}
+
+	searchSource := doc.Get("attributes.kibanaSavedObjectMeta.searchSourceJSON")
+	if searchSource.IsStr() {
+		if parsed, err := objx.FromJSON(searchSource.Str()); err == nil {
+			doc.Set("attributes.kibanaSavedObjectMeta.searchSourceJSON", parsed)
+		}
+	}
+
+	/* these transformations from the original script facilitate the vis_tsvb_aggs and other TSVB-related runtime fields
+		TODO - implement these or convert the
+		vis_tsvb_aggs and any other necessary runtime fields to Go
+	if (
+	    doc?.attributes?.visState?.params?.filter &&
+	    typeof doc.attributes.visState.params.filter !== "string"
+	  ) {
+		  console.log("ENCOUNTERED STRINGIFIED visState.params.filter STATE", path)
+	    doc.attributes.visState.params.filter = JSON.stringify(
+	      doc.attributes.visState.params.filter
+	    );
+	  }
+	  if (
+	    doc?.attributes?.visState?.params?.series &&
+	    Array.isArray(doc.attributes.visState.params.series)
+	  ) {
+		  console.log("ENCOUNTERED STRINGIFIED visState.params.series STATE", path)
+	    doc.attributes.visState.params.series =
+	      doc.attributes.visState.params.series.map((s) => ({
+	        ...s,
+	        filter: JSON.stringify(s.filter),
+	      }));
+	  }
+	*/
+
+	return doc, nil
+}
+
+// Report information about a visualization saved object (unmarshalled JSON)
+// Supports maps, saved searches, Lens, Vega, and legacy visualizations
+func DescribeVisualizationSavedObject(doc map[string]interface{}) (VisDesc, error) {
+	doc = objx.New(doc)
+	deserializeSubPaths(doc)
+
+	soType := doc["type"].(string)
```

Comment:
This will panic if `doc["type"]` is not a string, consider adding a check:
```suggestion
	soType, ok := doc["type"].(string)
	if !ok {
	    return VisDesc{}, errors.New("`type` in visualization is not present or is not a string")
	}
```